File: sdk/prismer-cloud/adapters/hermes/src/prismer_adapter_hermes/event_builder.py
"""
event_builder.py — Pure helper functions that construct PARA event dicts.

Each make_Xxx() returns a plain dict shaped per PARA spec §4.3.
No validation or side effects — callers are responsible for validation before emit.
Mirror of @prismer/adapters-core builders, Python edition.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional


# ─── Lifecycle family ──────────────────────────────────────────────────────────


def make_agent_register(
    agent_id: str,
    adapter: str,
    version: str,
    tiers_supported: List[int],
    capability_tags: List[str],
    workspace: str,
    workspace_group: Optional[str] = None,
) -> dict:
    agent: Dict[str, Any] = {
        "id": agent_id,
        "adapter": adapter,
        "version": version,
        "tiersSupported": tiers_supported,
        "capabilityTags": capability_tags,
        "workspace": workspace,
    }
    if workspace_group is not None:
        agent["workspaceGroup"] = workspace_group
    return {"type": "agent.register", "agent": agent}


def make_session_started(
    session_id: str,
    scope: str,
    parent_session_id: Optional[str] = None,
) -> dict:
    evt: Dict[str, Any] = {
        "type": "agent.session.started",
        "sessionId": session_id,
        "scope": scope,
    }
    if parent_session_id is not None:
        evt["parentSessionId"] = parent_session_id
    return evt


def make_session_ended(
    session_id: str,
    reason: str = "stop",
) -> dict:
    return {
        "type": "agent.session.ended",
        "sessionId": session_id,
        "reason": reason,
    }


def make_session_reset(
    session_id: str,
    reason: str = "reset",
) -> dict:
    return {
        "type": "agent.session.reset",
        "sessionId": session_id,
        "reason": reason,
    }


# ─── Turn / LLM family ─────────────────────────────────────────────────────────


def make_prompt_submit(
    session_id: str,
    prompt: str,
    source: str = "user",
) -> dict:
    return {
        "type": "agent.prompt.submit",
        "sessionId": session_id,
        "prompt": prompt,
        "source": source,
    }


def make_llm_pre(
    session_id: str,
    model: str,
    conversation_length: int,
    is_first_turn: bool,
) -> dict:
    return {
        "type": "agent.llm.pre",
        "sessionId": session_id,
        "model": model,
        "conversationLength": conversation_length,
        "isFirstTurn": is_first_turn,
    }


def make_llm_post(
    session_id: str,
    tokens_used: int,
    stop_reason: str,
) -> dict:
    return {
        "type": "agent.llm.post",
        "sessionId": session_id,
        "tokensUsed": tokens_used,
        "stopReason": stop_reason,
    }


def make_turn_step(
    session_id: str,
    iteration: int,
    tool_names: Optional[List[str]] = None,
) -> dict:
    return {
        "type": "agent.turn.step",
        "sessionId": session_id,
        "iteration": iteration,
        "toolNames": tool_names or [],
    }


def make_turn_end(
    session_id: str,
    last_assistant_message: Optional[str] = None,
) -> dict:
    evt: Dict[str, Any] = {
        "type": "agent.turn.end",
        "sessionId": session_id,
    }
    if last_assistant_message is not None:
        evt["lastAssistantMessage"] = last_assistant_message
    return evt


# ─── Tool family ───────────────────────────────────────────────────────────────


def make_tool_pre(
    call_id: str,
    tool: str,
    args: Any,
    risk_tag: Optional[str] = None,
) -> dict:
    evt: Dict[str, Any] = {
        "type": "agent.tool.pre",
        "callId": call_id,
        "tool": tool,
        "args": args,
    }
    if risk_tag is not None:
        evt["riskTag"] = risk_tag
    return evt


def make_tool_post(
    call_id: str,
    ok: bool,
    duration_ms: float,
    summary: str,
    updated_mcp_tool_output: Any = None,
) -> dict:
    evt: Dict[str, Any] = {
        "type": "agent.tool.post",
        "callId": call_id,
        "ok": ok,
        "durationMs": duration_ms,
        "summary": summary,
    }
    if updated_mcp_tool_output is not None:
        evt["updatedMCPToolOutput"] = updated_mcp_tool_output
    return evt


def make_tool_failure(
    call_id: str,
    error: str,
    signal_pattern: Optional[str] = None,
    is_interrupt: Optional[bool] = None,
) -> dict:
    evt: Dict[str, Any] = {
        "type": "agent.tool.failure",
        "callId": call_id,
        "error": error,
    }
    if signal_pattern is not None:
        evt["signalPattern"] = signal_pattern
    if is_interrupt is not None:
        evt["isInterrupt"] = is_interrupt
    return evt


# ─── Command family ────────────────────────────────────────────────────────────


def make_command(
    command: str,
    command_kind: str = "other",
    args: Any = None,
    source: Optional[str] = None,
) -> dict:
    evt: Dict[str, Any] = {
        "type": "agent.command",
        "command": command,
        "commandKind": command_kind,
    }
    if args is not None:
        evt["args"] = args
    if source is not None:
        evt["source"] = source
    return evt


# ─── Delegation family ─────────────────────────────────────────────────────────


def make_subagent_ended(
    parent_session_id: Optional[str],
    subagent_type: Optional[str],
    summary: Optional[str],
    status: str,
    duration_ms: float,
) -> dict:
    evt: Dict[str, Any] = {
        "type": "agent.subagent.ended",
        "status": status,
        "durationMs": duration_ms,
    }
    # Hermes delegate_tool passes parent_session_id=getattr(..., "session_id", None),
    # which may be None when the parent has no session_id. Omit rather than
    # emit parentSessionId="" so downstream can distinguish absence from empty.
    if parent_session_id:
        evt["parentSessionId"] = parent_session_id
    if subagent_type is not None:
        evt["subagentType"] = subagent_type
    if summary is not None:
        evt["summary"] = summary
    return evt


# ─── State family ──────────────────────────────────────────────────────────────


def make_agent_state(status: str) -> dict:
    return {"type": "agent.state", "status": status}

File: sdk/prismer-cloud/adapters-core/src/event-builder.ts
/**
 * event-builder.ts — Pure helper functions that assemble PARA events.
 *
 * Each constructor:
 *   1. Accepts typed params (no `any`)
 *   2. Validates the assembled event via ParaEventSchema.parse() — throws
 *      ZodError on invalid input so adapter bugs surface immediately
 *   3. Returns the validated ParaEvent (discriminated union member)
 *
 * One constructor per event family that adapters commonly emit (~20 events).
 * Rare events (elicitation.result, channel.transcribed, skill.proposed) are
 * intentionally omitted — consumers construct literals and call
 * ParaEventSchema.parse() directly.
 */

import type {
  AgentDescriptor,
  ParaEvent,
  SkillSource,
} from '@prismer/wire';
import { ParaEventSchema } from '@prismer/wire';

// ─── Helper: validate and narrow ──────────────────────────────────────────

function build(raw: unknown): ParaEvent {
  return ParaEventSchema.parse(raw);
}

// ─── Lifecycle family ─────────────────────────────────────────────────────

/** agent.register — emitted once at adapter startup. */
export function makeRegisterEvent(agent: AgentDescriptor): ParaEvent {
  return build({ type: 'agent.register', agent });
}

/** agent.session.started — emitted when a new session begins. */
export function makeSessionStarted(params: {
  sessionId: string;
  scope: string;
  parentSessionId?: string;
}): ParaEvent {
  return build({ type: 'agent.session.started', ...params });
}

/** agent.session.reset — emitted on /clear, /reset, or compact-triggered reset. */
export function makeSessionReset(params: {
  sessionId: string;
  reason: 'new' | 'reset' | 'clear' | 'compact';
}): ParaEvent {
  return build({ type: 'agent.session.reset', ...params });
}

/** agent.session.ended — emitted when the session terminates. */
export function makeSessionEnded(params: {
  sessionId: string;
  reason: 'stop' | 'crash' | 'quota' | 'logout' | 'other';
}): ParaEvent {
  return build({ type: 'agent.session.ended', ...params });
}

/** agent.state — emitted whenever the agent transitions between idle/thinking/tool/etc. */
export function makeAgentState(
  status: 'idle' | 'thinking' | 'tool' | 'awaiting_approval' | 'error',
): ParaEvent {
  return build({ type: 'agent.state', status });
}

// ─── Turn / LLM family ────────────────────────────────────────────────────

/** agent.prompt.submit — emitted when a user/remote/subagent prompt enters the turn. */
export function makePromptSubmit(params: {
  sessionId: string;
  prompt: string;
  source: 'user' | 'remote' | 'subagent';
}): ParaEvent {
  return build({ type: 'agent.prompt.submit', ...params });
}

/** agent.llm.pre — emitted just before sending a request to the LLM. */
export function makeLlmPre(params: {
  sessionId: string;
  model: string;
  conversationLength: number;
  isFirstTurn: boolean;
}): ParaEvent {
  return build({ type: 'agent.llm.pre', ...params });
}

/** agent.llm.post — emitted after the LLM responds. */
export function makeLlmPost(params: {
  sessionId: string;
  tokensUsed: number;
  stopReason: string;
}): ParaEvent {
  return build({ type: 'agent.llm.post', ...params });
}

/** agent.turn.step — emitted after each tool call in a turn. */
export function makeTurnStep(params: {
  sessionId: string;
  iteration: number;
  toolNames: string[];
}): ParaEvent {
  return build({ type: 'agent.turn.step', ...params });
}

/** agent.turn.end — emitted when the agent finishes a full turn. */
export function makeTurnEnd(params: {
  sessionId: string;
  lastAssistantMessage?: string;
}): ParaEvent {
  return build({ type: 'agent.turn.end', ...params });
}

/** agent.turn.failure — emitted when a turn fails with a typed error. */
export function makeTurnFailure(params: {
  sessionId: string;
  errorType: 'rate_limit' | 'auth' | 'billing' | 'invalid' | 'server' | 'max_tokens' | 'unknown';
  errorMessage: string;
}): ParaEvent {
  return build({ type: 'agent.turn.failure', ...params });
}

// ─── Tool family ──────────────────────────────────────────────────────────

/** agent.tool.pre — emitted before a tool is executed. */
export function makeToolPre(params: {
  callId: string;
  tool: string;
  args: unknown;
  riskTag?: 'low' | 'mid' | 'high';
}): ParaEvent {
  return build({ type: 'agent.tool.pre', ...params });
}

/** agent.tool.post — emitted after a tool completes successfully. */
export function makeToolPost(params: {
  callId: string;
  ok: boolean;
  durationMs: number;
  summary: string;
  updatedMCPToolOutput?: unknown;
}): ParaEvent {
  return build({ type: 'agent.tool.post', ...params });
}

/** agent.tool.failure — emitted when a tool invocation fails or is interrupted. */
export function makeToolFailure(params: {
  callId: string;
  error: string;
  signalPattern?: string;
  isInterrupt?: boolean;
}): ParaEvent {
  return build({ type: 'agent.tool.failure', ...params });
}

// ─── Permission family ────────────────────────────────────────────────────

/** agent.approval.request — emitted when the agent needs user/remote approval. */
export function makeApprovalRequest(params: {
  callId: string;
  prompt: string;
  ttlMs: number;
  permissionSuggestions?: unknown[];
}): ParaEvent {
  return build({ type: 'agent.approval.request', ...params });
}

/** agent.approval.result — emitted after an approval decision is received. */
export function makeApprovalResult(params: {
  callId: string;
  decision: 'allow' | 'deny' | 'ask' | 'defer';
  by: 'local' | 'remote';
  updatedInput?: unknown;
  updatedPermissions?: unknown[];
}): ParaEvent {
  return build({ type: 'agent.approval.result', ...params });
}

// ─── Task family ──────────────────────────────────────────────────────────

/** agent.task.created — emitted when the agent creates a task in the IM system. */
export function makeTaskCreated(params: {
  taskId: string;
  subject: string;
  description?: string;
  teammateName?: string;
  teamName?: string;
}): ParaEvent {
  return build({ type: 'agent.task.created', ...params });
}

/** agent.task.completed — emitted when a task finishes. */
export function makeTaskCompleted(params: {
  taskId: string;
  subject: string;
  status: 'completed' | 'failed' | 'cancelled';
}): ParaEvent {
  return build({ type: 'agent.task.completed', ...params });
}

// ─── Memory / Compact family ──────────────────────────────────────────────

/** agent.compact.pre — emitted before context compaction. */
export function makeCompactPre(params: {
  sessionId: string;
  trigger: 'manual' | 'auto';
  messageCount: number;
  tokenCount: number;
}): ParaEvent {
  return build({ type: 'agent.compact.pre', ...params });
}

/** agent.compact.post — emitted after context compaction completes. */
export function makeCompactPost(params: {
  sessionId: string;
  compactedCount: number;
  tokensBefore: number;
  tokensAfter: number;
}): ParaEvent {
  return build({ type: 'agent.compact.post', ...params });
}

/** agent.bootstrap.injected — emitted when bootstrap files are injected into context. */
export function makeBootstrapInjected(params: {
  bootstrapFiles: string[];
  agentId: string;
}): ParaEvent {
  return build({ type: 'agent.bootstrap.injected', ...params });
}

// ─── Skill family ─────────────────────────────────────────────────────────

/** agent.skill.activated — emitted when a skill is invoked/matched. */
export function makeSkillActivated(params: {
  skillName: string;
  source: SkillSource;
  trigger: 'user-invoke' | 'model-invoke' | 'auto-match';
  args?: string;
}): ParaEvent {
  return build({ type: 'agent.skill.activated', ...params });
}

/** agent.skill.deactivated — emitted when a skill is dropped (compaction, session-end, explicit). */
export function makeSkillDeactivated(params: {
  skillName: string;
  reason: 'compaction-drop' | 'session-end' | 'explicit';
}): ParaEvent {
  return build({ type: 'agent.skill.deactivated', ...params });
}

/** agent.skill.proposed — emitted when agent creates a skill draft (e.g., via skill_manage tool). */
export function makeSkillProposed(params: {
  draftPath: string;
  name: string;
  description: string;
  author: 'agent' | 'user';
}): ParaEvent {
  return build({ type: 'agent.skill.proposed', ...params });
}

/** agent.skill.installed — emitted when a skill is installed from a registry. */
export function makeSkillInstalled(params: {
  skillName: string;
  source: SkillSource;
  version?: string;
  sha256: string;
}): ParaEvent {
  return build({ type: 'agent.skill.installed', ...params });
}

/** agent.skill.uninstalled — emitted when a skill is removed. */
export function makeSkillUninstalled(params: {
  skillName: string;
}): ParaEvent {
  return build({ type: 'agent.skill.uninstalled', ...params });
}

// ─── Message I/O family ──────────────────────────────────────────

/** agent.message — emitted for every message (user/agent/system) in conversation. */
export function makeAgentMessage(params: {
  role: 'user' | 'agent' | 'system';
  content: string;
  ts: number;
}): ParaEvent {
  return build({ type: 'agent.message', ...params });
}

/** agent.channel.inbound — emitted when a message arrives from external channel (Telegram, Discord, etc.). */
export function makeChannelInbound(params: {
  from: string;
  content: string;
  channelId: string;
  metadata?: Record<string, unknown>;
}): ParaEvent {
  return build({ type: 'agent.channel.inbound', ...params });
}

/** agent.channel.outbound.sent — emitted after sending to external channel. */
export function makeChannelOutboundSent(params: {
  to: string;
  content: string;
  channelId: string;
  success: boolean;
}): ParaEvent {
  return build({ type: 'agent.channel.outbound.sent', ...params });
}

/** agent.channel.transcribed — emitted after audio transcription completes. */
export function makeChannelTranscribed(params: {
  transcript: string;
  from: string;
  channelId: string;
  mediaPath?: string;
}): ParaEvent {
  return build({ type: 'agent.channel.transcribed', ...params });
}

/** agent.channel.preprocessed — emitted after media/enrichment preprocessing. */
export function makeChannelPreprocessed(params: {
  bodyForAgent: string;
  from: string;
  channelId: string;
}): ParaEvent {
  return build({ type: 'agent.channel.preprocessed', ...params });
}

// ─── Environment family ───────────────────────────────────────────────────

/** agent.fs.op — emitted for file system operations (read/write/delete/etc). */
export function makeFsOp(params: {
  sessionId: string;
  operation: 'read' | 'write' | 'delete' | 'create' | 'rename' | 'move';
  path: string;
  success: boolean;
  error?: string;
}): ParaEvent {
  return build({ type: 'agent.fs.op', ...params });
}

/** agent.file.watched — emitted when a watched file is modified. */
export function makeFileWatched(params: {
  sessionId: string;
  path: string;
  event: 'created' | 'modified' | 'deleted';
}): ParaEvent {
  return build({ type: 'agent.file.watched', ...params });
}

/** agent.cwd.changed — emitted when current working directory changes. */
export function makeCwdChanged(params: {
  sessionId: string;
  oldCwd: string;
  newCwd: string;
}): ParaEvent {
  return build({ type: 'agent.cwd.changed', ...params });
}

/** agent.config.changed — emitted when configuration is modified. */
export function makeConfigChanged(params: {
  sessionId: string;
  configPath: string;
  keysChanged: string[];
}): ParaEvent {
  return build({ type: 'agent.config.changed', ...params });
}

/** agent.worktree.created — emitted when a git worktree is created. */
export function makeWorktreeCreated(params: {
  sessionId: string;
  worktreePath: string;
  baseBranch: string;
}): ParaEvent {
  return build({ type: 'agent.worktree.created', ...params });
}

/** agent.worktree.removed — emitted when a git worktree is removed. */
export function makeWorktreeRemoved(params: {
  sessionId: string;
  worktreePath: string;
}): ParaEvent {
  return build({ type: 'agent.worktree.removed', ...params });
}
